cli/internal/cache/cache_fs.go

```diff
@@ -89,7 +89,7 @@ func (f *fsCache) Put(target, hash string, duration int, files []string) error {
 						return fmt.Errorf("error ensuring directory file from cache: %w", err)
 					}
 
-					if err := fs.CopyOrLinkFile(file, filepath.Join(f.cacheDirectory, hash, file), fromInfo.Mode(), fs.DirPermissions, false, false); err != nil {
+					if err := fs.CopyOrLinkFile(&fs.LstatCachedFile{Path: fs.AbsolutePath(file), Info: fromInfo}, filepath.Join(f.cacheDirectory, hash, file), false, false); err != nil {
```

Comment:
Re: `fs.AbsolutePath`, please don't cast to it. Use one of `CheckedToAbsolutePath` if you don't know if it's an absolute path, `UnsafeToAbsolutePath` if you are certain that it's an absolute path but we haven't migrated the calling code to use `AbsolutePath` yet, and `AbsolutePathFromUpstream` if the value is coming from some library function that you are certain is providing absolute paths.

This helps us find all the places that need to be migrated, as well as communicates what APIs have been investigated already. I expect this applies to anything from `turbopath` as well.

Reply:
Indeed, the `turbopath` stuff is following the same rules. (Good catch on this cast, I noticed a different one.)

Some notes on `turbopath`:
- `AbsoluteSystemPath` models the same thing as `turbofs.AbsolutePath`. There will be a migration PR which should hopefully "just work." I just tried to do it in an automated fashion and everything broke terribly so I need to figure out what I did wrong.
- At some point there may be a bonus migration where we make a `AnchoredXxxxPath` type that is `io/fs`-friendly, but that's not today.

Ordering for things:
1. Land the daemon code which extensively uses `AbsolutePath` so that we don't have to do it twice. (Friday US?)
2. Migrate `turbofs.AbsolutePath` to `turbopath.AbsoluteSystemPath`. (Monday HK/US?)
3. Delete anything unused that remains. (Tuesday HK/US?)

Reply:
@gsoltis I want to make sure we're on the same page with terminology. 

I interpret absolute path as a path from system root. For our purposes, is that valid here? Or is our notion of an absolute path (`fs.AbsolutePath`) a path from the _monorepo_ root.

I'm asking because If the former, _most_ (if not all) of the paths in this PR are relative and will need to be converted before using one of `CheckedToAbsolutePath`, `UnsafeToAbsolutePath`, or `AbsolutePathFromUpstream`

Reply:
`fs.AbsolutePath` is semantically equivalent to `turbopath.AbsoluteSystemPath` and represents an absolute path in the native format of the particular OS. So `<Volume>:\some\path` or `/some/path` as appropriate. 

The motivation for all this work separating out different kinds of paths is that they show up in our user input as strings, and all of the APIs for interacting with the filesystem and for interacting with potentially-shared cache artifacts also operate on strings, and it is tempting to naively assume one can be passed to the other. This has been a source of bugs, as well as my suspicion that the `--cwd` flag doesn't _really_ work as intended. 

The solution we've landed on is attempting to identify the specific kinds of paths we care about internally and being strict about what we're working with at any given point. The two important ones are `fs.AbsolutePath/turbopath.AbsoluteSystemPath` for anything that touches the filesystem, and `turbopath.AnchoredUnixPath` for anything that is potentially shared.

We've been migrating gradually, so some parts of the code are already using `fs.AbsolutePath` for filesystem access. The `turbopath` stuff is new, and should help us further clarify paths especially around the cache. 

There's a judgement call in how much to migrate at any one point. You will likely at some point end up doing `UnsafeTo<SomePathType>` with a comment explaining that you've verified that the `string` is of the correct type. Likewise, if you need to interact with some part of the system that hasn't migrated yet, you can do `ToStringDuringMigration` (you may need to add this, depending what kind of type you already have). This lets us easily find the boundaries of what has and hasn't been migrated.

If it's not there already, you should feel free to thread `Cofing.Cwd` into whatever code you're working with. It is the `fs.AbsolutePath` to the root of the monorepo. You can then use your relative paths to get absolute paths for filesystem interactions.

Since it sounds like the transition from `fs.AbsolutePath` to `turbopath.AbsoluteSystemPath` is going to be an automated one, I would go with the `fs.AbsolutePath` for now, since it has the filesystem operations mostly implemented already. But I would except the migration to happen in short order.